ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Client.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+module Ouroboros.Network.Protocol.BlockFetch.Client where
+
+import Protocol.Core
+import Pipes (Producer)
+import qualified Pipes
+
+import Ouroboros.Network.Protocol.BlockFetch.Type
+
+{-------------------------------------------------------------------------------
+  Client stream of @'BlockFetchClientProtocol'@ protocol
+-------------------------------------------------------------------------------}
+
+data BlockFetchClientSender range m a where
+
+  -- | Send a request for a range of blocks.
+  --
+  SendBlockRequestMsg
+    :: range
+    -> m (BlockFetchClientSender range m a)
+    -> BlockFetchClientSender range m a
+
+  -- | The client decided to end the protocol.
+  --
+  SendMsgDone
+    :: a
+    -> BlockFetchClientSender range m a
+
+blockFetchClientSenderFromProducer
+  :: Monad m
+  => Producer range m a
+  -> m (BlockFetchClientSender range m a)
+blockFetchClientSenderFromProducer producer = Pipes.next producer >>= \nxt -> case nxt of
+  Left a                   -> return $ SendMsgDone a
+  Right (range, producer') -> return $ SendBlockRequestMsg range (blockFetchClientSenderFromProducer producer')
+
+blockFetchClientSenderStream
+  :: ( Functor m )
+  => BlockFetchClientSender range m a
+  -> Peer BlockFetchClientProtocol
+      (BlockRequestClientMessage range)
+      (Yielding StClientIdle)
+      (Finished StClientDone)
+      m a
+blockFetchClientSenderStream (SendBlockRequestMsg range next) =
+  part (MessageRequestRange range) $
+    lift (blockFetchClientSenderStream <$> next)
+blockFetchClientSenderStream (SendMsgDone a) = out MessageDone (done a)
+
+{-------------------------------------------------------------------------------
+  Client stream of @'BlockFetchServerProtocol'@ protocol
+-------------------------------------------------------------------------------}
+
+-- | Handlers for server responses>
+--
+data BlockFetchClientReceiver block m a =
+     BlockFetchClientReceiver {
+
+       -- | The server has started streaming blocks.
+       --
+       recvMsgStartBatch :: m (BlockFetchClientReceiveBlocks block m a),
+
+       -- | The block fetch request failed because a requested range was not on
+       -- a producer's chain.
+       --
+       recvMsgNoBlocks   :: m (BlockFetchClientReceiver block m a),
+
+       -- | The server terminated the protocol.
+       --
+       recvMsgDoneClient :: a
+     }
+
+-- | Block download handlers.
+--
+data BlockFetchClientReceiveBlocks block m a = 
+     BlockFetchClientReceiveBlocks {
+       recvMsgBlock       :: block -> m (BlockFetchClientReceiveBlocks block m a),
+       recvMsgBatchDone   :: m (BlockFetchClientReceiver block m a),
+       recvMsgServerError :: m (BlockFetchClientReceiver block m a)
+     }
+
+-- | Construct @'Peer' 'BlockFetchServerProtocol'@ from
+-- @'BlockFetchClientReceiver'@.
+--
+blockFetchClientReceiverStream
+  :: forall m block a.
+     ( Functor m )
+  => BlockFetchClientReceiver block m a
```

Comment:
Can we make one of these from a pipes consumer? Dual to the `blockFetchClientSenderFromProducer`?

Reply:
I'd imagine we ought to be able to, or else there's some problem with these definitions.

Reply:
If it's dual then it should be `toConsumer`,but then a bit useless when using
a `Peer`.  The thing is that we have `BlockFetchClientReceiveBlocks block
m a -> Peer ...` so `fromConsumer` fits better.  But without a `Peer` it makes
all sense:

* `blockFetchClientSenderFromProducer :: Producer range m a -> BlockFetchClientSender range m a`
* `blockFetchServerReceiverPipe       :: (range -> Producer' block m a) -> BlockFetchServerReceiver range m Void -> Pipe range block m a`
* `blockFetchServerSenderToProducer   :: BlockFetchServerSender block m a -> Producer' block m a`
* `blockFetchServerReceiverToConsumer   :: BlockFetchServerReceiver block m a -> Consumer block m a`

Then one can use `connectThroughQueue` to create `BlockFetchServerReceiver` and
`BlockFetchServerSender` and use those functions to run them as a pipe.
I implemented first three, and I am pretty sure the third one can be done too.
The reason I removed them is that I just didn't see a reason to keep them
around, but it did convienced me that the approach makes sense from `Pipes`
perspective.  I initially thought about using this in tests, but then I could
just make pure computation to check if the result of `direct` and `connect`
functions is the correct one.

Reply:
As for `blockFetchServerFromConsumer :: Consumer block m a -> BlockFetchServerReceiver block m a` it will be necessarily a partial function; since the `BlockFetchServerReceiver` can handle three cases: start a batch (which is consumer like), handle `MessageNoBlocks` and `MessageServerDone`, the consumer will only know how to fill one of two of those (depending on its constructor).  If we treat `BlocKFetchServerReceiver` (a linear disjunction would fit better).

Actually similar thing for a `toConsumer`, since `BlockFetchClientReceiver block m a` can deal with three cases a consumer can only deal with one:
```
blockFetchClientReceiverToConsumer
  :: forall block m a. Monad m
  => BlockFetchClientReceiver block m a
  -> Consumer block m a
blockFetchClientReceiverToConsumer BlockFetchClientReceiver{recvMsgStartBatch} = join $ Pipes.lift $ do
    receiver <- recvMsgStartBatch
    return (receiverToConsumer receiver)
 where
  receiverToConsumer :: BlockFetchClientReceiveBlocks block m a -> Consumer block m a
  receiverToConsumer BlockFetchClientReceiveBlocks {recvMsgBlock} = Pipes.Request () $ \block -> join $ Pipes.lift $ recvMsgBlock block >>= return . receiverToConsumer
```
The problem is that a consumer does not now how to drive through the graph of all possible transitions, one has to choose a single path and for that path one can produce a consumer.  This could be done with the `Path` module, but apart as being a fun exercise I don't see a good reason to do that.